dtm0/dtm0.h

```diff
@@ -0,0 +1,116 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+#pragma once
+
+#ifndef __MOTR_DTM0_DTM0_H__
+#define __MOTR_DTM0_DTM0_H__
+
+/**
+ * @page dtm0s-fspec The dtm0 service (DTM0S)
+ *
+ * DTM0 will reuse existing cas service and cas fops/foms to deliver dtx related
+ * information.
+ *
+ * Cas fop will aggregate additional data related to dtx, txr {list of
+ * participants, versions, ...}. It creates a tx descriptor containing this
+ * information and embeds it in cas_op as a dtm0 transaction payload.
+ *
+ * The interface of the service for a user is a format of request/reply FOPs.
+ * DTM0 service registers following FOP types during initialisation:
+ * - @ref dtm0_req_fop_fopt
+ * - @ref dtm0_rep_fop_fopt
+ *
+ * DTM supports three kinds of FOP/FOM operations.
+ * - EXECUTED FOP/FOM: when a cas fop is being executed, cas service will check
+ *   for the existence of the dtm transaction information. If it is present, cas
+ *   will log the tx descriptor locally on each of the dtm participants.  dtm0
+ *   log will be updated inside cas fom (ad property of tx). The txr will be
+ *   executed just before tx_close phase of the fom inside fom_fol_rec_add().
+ * - REDO FOP/FOM: When a participant service restarts after a transient failure
+ *   the other existing participants sends REDO requests to the recovering
+ *   participant.
+ * - PERSISTENT FOP/FOM: When the transaction get logged, DTM sends a PERSISTENT
+ *   FOP to the other participants and to the originator indicating that the
+ *   change caused by the Cas operation has become permanent.
+ *
+ * @subsection dtm0-state State Specification
+ *
+ * @verbatim
+ *
+ *			    |
+ *			    V
+ *			FOPH_INIT
+ *			    |
+ *			    V
+ *		    [generic phases]
+ *			    |
+ *			    V
+ *			 TXN_INIT
+ * 			    |
+ *			    V
+ *			 TXN_OPEN
+ *			    |
+ *			    V
+ *		    UPDATE DTM0 LOG
+ *			    |
+ *			    V
+ *			 TXN_CLOSE
+ *			    |
+ *			    V
+ *			FOPH_FINI
+ *
+ * @verbatim
+ *
+ * The primary datastructure that describes a user operation and all the
+ * associated data for the request is dtms0_op. This structure has a member
+ * field that carries an operation code. The operation code specifies what kind
+ * of an operation this is. It can be one of:
+ * - DT_REQ: for a dtm0 request, or
+ * - DT_REDO: for a dtm0 redo-request operation
+ * Each dtm0_op also carries a dtm_msg type which tells what kind of message it
+ * is carrying.  A dtms0_op containing a request must specify a message type of
+ * either DMT_EXECUTE_DTX,indicating that this request contains an operation to
+ * be executed, a m0_dtms0_op containing a reply operation should contain either
+ * a DMT_EXECUTED or a DMT_PERSISTENT message.
+ */
+
+/**
+ * @{
```

Comment:
We need to add an API that allows us to initialize a DTM0 service instance (DTM0 service + DTM0 log + clock source). It will be used inside motr/client_init.c on the originator side. At this moment it is possible to initialize DTM0 service only on the server side (motr/init.c).

Reply:
Just to clarify: it was just a general comment, we don't have to address it in the scope of this patch. It looks like it will be covered (at least partially) by this patch (https://github.com/Seagate/cortx-motr/pull/382). There are `m0_dtm__client_service_start` and `m0_reqh_service_lookup`.

Reply:
OK. Will review m0_dtm__client_service_start routine from the above patch and then discuss what is missing and what needs to be added.